clang/lib/Lex/PPMacroExpansion.cpp

```diff
@@ -1673,6 +1673,7 @@ void Preprocessor::ExpandBuiltinMacro(Token &Tok) {
         return II && HasExtension(*this, II->getName());
       });
   } else if (II == Ident__has_builtin) {
+    EvaluatingHasBuiltinMacro = true;
```

Comment:
I'm not particularly proud of this workaround, and if reviewers think that we should do this in some other way (e.g. enable backtracking for every `__has_builtin` use, and look behind when we see a revertible type trait identifier), I'm open for suggestions.